### ISSUE TITLE
mso_schema_site_anp_epg_domain - adding domain to EPG fails initially

### ISSUE THREAD

**Issue by sacascio** [CiscoDevNet/ansible-mso#38]
When adding a domain profile to an existing EPG in an existing schema/template, the ansible script fails indicating that the EPG is not found.  After adding the domain profile to the EPG manually in MSO, re-executing the ansible script is successful.  Details below.

ansible 2.9.3
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/scascio/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/scascio/venv_ansible_2_9_3_python_2_7_17/lib/python2.7/site-packages/ansible
  executable location = /home/scascio/venv_ansible_2_9_3_python_2_7_17/bin/ansible
  python version = 2.7.17 (default, Jun  5 2020, 12:02:28) [GCC 4.8.5 20150623 (Red Hat 4.8.5-39)]

Schema: SalsSchema
Tenant: HR
AP: NewAP
EPG: AnotherEPG
VRF: NewVRF
BD: NewBD

1) Run Ansible to add a domain profile to an EPG which has no domain profiles associated to it:

 - name: Add domain to EPG
      mso_schema_site_anp_epg_domain:
       hostname: "192.168.1.138"
       username: "admin"
       password: "We!come2msc!"
       validate_certs: no
       schema: 'SalsSchema'
       site: 'DC1'
       template: 'NewTemplate'
       anp: 'NewAP'
       epg: ‘AnotherEPG'
       domain_association_type: 'physicalDomain'
       domain_profile: 'phys2'
       deployment_immediacy: 'immediate'
       resolution_immediacy: 'immediate'
       state: present
      register: result
      delegate_to: localhost
    - debug: msg="{{result}}”

TASK [Add domain to EPG] *************************************************************************************************************************
fatal: [localhost -> localhost]: FAILED! => {
    "changed": false, 
    "current": {}
}

MSG:

Provided epg 'AnotherEPG' does not exist. Existing epgs: {u'externalEpgs': [], u'templateName': u'NewTemplate', u'contracts': [], u'anps': [{u'anpRef': u'/schemas/5edea58f120000b10c2dc766/templates/NewTemplate/anps/NewAP', u'epgs': [{u'subnets': [], u'domainAssociations': [{u'dn': u'uni/phys-phys', u'deployImmediacy': u'immediate', u'resolutionImmediacy': u'immediate', u'domainType': u'physicalDomain'}], u'uSegAttrs': [], u'staticPorts': [], u'staticLeafs': [], u'selectors': [], u'epgRef': u'/schemas/5edea58f120000b10c2dc766/templates/NewTemplate/anps/NewAP/epgs/NewEPG'}]}], u'siteId': u'5ede69af10000024017a4cfc', u'vrfs': [], u'bds': [], u'intersiteL3outs': [], u'serviceGraphs': []} epgref /schemas/5edea58f120000b10c2dc766/templates/NewTemplate/anps/NewAP/epgs/AnotherEPG

2) Result from REST QUERY of EPG object in MSO:

{
                                    "name": "AnotherEPG",
                                    "displayName": "AnotherEPG",
                                    "epgRef": "/schemas/5edea58f120000b10c2dc766/templates/NewTemplate/anps/NewAP/epgs/AnotherEPG",
                                    "contractRelationships": [],
                                    "subnets": [],
                                    "uSegEpg": false,
                                    "uSegAttrs": [],
                                    "intraEpg": "unenforced",
                                    "proxyArp": false,
                                    "preferredGroup": false,
                                    "bdRef": "/schemas/5edea58f120000b10c2dc766/templates/NewTemplate/bds/NewBD",
                                    "vrfRef": "",
                                    "selectors": [

3) Manually add ‘phys’ physicalDomain to AnotherEPG in MSO

4) Re-run ansible, it works:

TASK [Add domain to EPG] *************************************************************************************************************************
changed: [localhost -> localhost]

TASK [debug] *************************************************************************************************************************************
ok: [localhost] => {}

MSG:

{u'current': {u'dn': u'uni/phys-phys2', u'deploymentImmediacy': u'immediate', u'resolutionImmediacy': u'immediate', u'domainType': u'physicalDomain'}, 'failed': False, u'changed': True}

PLAY RECAP ***************************************************************************************************************************************
localhost                  : ok=2    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0

**Comment by lhercot** [comment]
@sacascio, this looks like a duplicate of #25.

Can you try using the mso_schema_site_anp and mso_schema_site_anp_epg before the mso_schema_site_anp_epg_domain to create the required structure?

**Comment by sacascio** [comment]
I tried that, but it looks like it was failing to create the ANP.  Only after I created the ANP manually in the UI did the task work to add the ANP.  
The EPG then created successfully, however, adding the domain still failed.  here is the full detail:


TASK [Add a new ANP] *********************************************************************************************************************************************************************************************
fatal: [localhost -> localhost]: FAILED! => {
    "changed": true, 
    "current": {
        "anpRef": {
            "anpName": "ANP0609", 
            "schemaId": "5edfda70140000f307030750", 
            "templateName": "Template0609"
        }
    }, 
    "data": [
        {
            "op": "add", 
            "path": "/sites/5ede69af10000024017a4cfc-Template0609/anps/-", 
            "value": {
                "anpRef": {
                    "anpName": "ANP0609", 
                    "schemaId": "5edfda70140000f307030750", 
                    "templateName": "Template0609"
                }
            }
        }
    ], 
    "info": {
        "body": "{\"code\":142,\"message\":\"Invalid Reference: Unable to find AP 'AnpRef(5edfda70140000f307030750,Template0609,ANP0609)' \"}", 
        "connection": "close", 
        "content-language": "en", 
        "content-length": "118", 
        "content-type": "application/json", 
        "date": "Tue, 09 Jun 2020 19:06:18 GMT", 
        "msg": "HTTP Error 400: Bad Request", 
        "server": "nginx", 
        "status": 400, 
        "strict-transport-security": "max-age=31536000; includeSubDomains", 
        "url": "https://192.168.1.140/api/v1/schemas/5edfda70140000f307030750"
    }, 
    "payload": {
        "code": 142, 
        "message": "Invalid Reference: Unable to find AP 'AnpRef(5edfda70140000f307030750,Template0609,ANP0609)' "
    }
}

MSG:

**MSO Error 142: Invalid Reference: Unable to find AP 'AnpRef(5edfda70140000f307030750,Template0609,ANP0609)'** 



2. Add ANP manually, re-run ansible-playbook

(venv_ansible_2_9_3_python_2_7_17) [scascio@vader ~]$ ansible-playbook do_all.yml 

PLAY [Add ANP, EPG and update physical domain] *******************************************************************************************************************************************************************

TASK [Add a new ANP] *********************************************************************************************************************************************************************************************
changed: [localhost -> localhost]

TASK [debug] *****************************************************************************************************************************************************************************************************
ok: [localhost] => {}

MSG:

{u'current': {u'anpRef': {u'schemaId': u'5edfda70140000f307030750', u'anpName': u'ANP0609', u'templateName': u'Template0609'}}, 'failed': False, u'changed': True}


TASK [Add a new EPG] *********************************************************************************************************************************************************************************************
changed: [localhost -> localhost]

TASK [Add domain to EPG] *****************************************************************************************************************************************************************************************
fatal: [localhost -> localhost]: FAILED! => {
    "changed": false, 
    "current": {}
}

MSG:

**Provided epg 'EPG0609' does not exist.** Existing epgs: {u'externalEpgs': [], u'templateName': u'Template0609', u'contracts': [], u'anps': [{u'anpRef': u'/schemas/5edfda70140000f307030750/templates/Template0609/anps/ANP0609', u'epgs': []}], u'siteId': u'5ede69af10000024017a4cfc', u'vrfs': [{u'regions': [], u'vrfRef': u'/schemas/5edfda70140000f307030750/templates/Template0609/vrfs/VRF0609'}], u'bds': [], u'intersiteL3outs': [], u'serviceGraphs': []} epgref /schemas/5edfda70140000f307030750/templates/Template0609/anps/ANP0609/epgs/EPG0609


PLAY RECAP *******************************************************************************************************************************************************************************************************
localhost                  : ok=3    changed=2    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0   

3. Add physical domain manually, then run ansible-playbook again to add a second physical domain - all good

venv_ansible_2_9_3_python_2_7_17) [scascio@vader ~]$ ansible-playbook do_all.yml 

PLAY [Add ANP, EPG and update physical domain] *******************************************************************************************************************************************************************

TASK [Add a new ANP] *********************************************************************************************************************************************************************************************
changed: [localhost -> localhost]

TASK [debug] *****************************************************************************************************************************************************************************************************
ok: [localhost] => {}

MSG:

{u'current': {u'anpRef': {u'schemaId': u'5edfda70140000f307030750', u'anpName': u'ANP0609', u'templateName': u'Template0609'}, u'epgs': [{u'subnets': [], u'domainAssociations': [{u'dn': u'uni/phys-phys', u'deployImmediacy': u'immediate', u'resolutionImmediacy': u'immediate', u'domainType': u'physicalDomain'}], u'epgRef': u'/schemas/5edfda70140000f307030750/templates/Template0609/anps/ANP0609/epgs/EPG0609', u'uSegAttrs': [], u'staticPorts': [], u'staticLeafs': [], u'selectors': []}]}, 'failed': False, u'changed': True}


TASK [Add a new EPG] *********************************************************************************************************************************************************************************************
ok: [localhost -> localhost]

TASK [Add domain to EPG] *****************************************************************************************************************************************************************************************
changed: [localhost -> localhost]

TASK [debug] *****************************************************************************************************************************************************************************************************
ok: [localhost] => {}

MSG:

{u'current': {u'dn': u'uni/phys-phys2', u'deploymentImmediacy': u'immediate', u'resolutionImmediacy': u'immediate', u'domainType': u'physicalDomain'}, 'failed': False, u'changed': True}


PLAY RECAP *******************************************************************************************************************************************************************************************************
localhost                  : ok=5    changed=2    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0

**Comment by lhercot** [comment]
@sacascio please update to the new v0.0.6 of the collection using: 
`ansible-galaxy collection install cisco.mso:0.0.6 --force`

This issue should be fixed in this version. Let us knwo and we will reopen the issue if needed.
